can/CAN.py

```diff
@@ -1,308 +1,379 @@
+    """
+
+    def __init__(self):
+        self.buffer = queue.Queue(0)
+
+    def on_message_received(self, msg):
+        self.buffer.put(msg)
+
+    def get_message(self, timeout=0.5):
+        """
+        Attempts to retrieve the latest message received by the instance. If no message is
+        available it blocks for given timeout or until a message is received (whichever
+        is shorter),
+
+        :param float timeout: The number of seconds to wait for a new message.
+        :return: the :class:`~can.Message` if there is one, or None if there is not.
+        """
+        try:
+            return self.buffer.get(block=True, timeout=timeout)
+        except queue.Empty:
+            return None
+
+
+class Logger(object):
+    """
+    Logs CAN messages to a file.
+
+    The format is determined from the file format which can be one of:
+      * .asc: :class:`can.ASCWriter`
+      * .csv: :class:`can.CSVWriter`
+      * .db: :class:`can.SqliteWriter`
+      * other: :class:`can.Printer`
+    """
+
+    @classmethod
+    def __new__(cls, other, filename):
+        if not filename:
+            return Printer()
+        elif filename.endswith(".asc"):
+            return ASCWriter(filename)
+        elif filename.endswith(".csv"):
+            return CSVWriter(filename)
+        elif filename.endswith(".db"):
+            return SqliteWriter(filename)
+        else:
+            return Printer(filename)
+
+
+class LogReader(object):
+    """
+    Replay logged CAN messages from a file.
+
+    The format is determined from the file format which can be one of:
+      * .asc
+      * .csv
+      * .db
+
+    Exposes a simple iterator interface, to use simply:
+
+        >>> for m in LogReader(my_file):
+        ...     print(m)
+
+    Note there are no time delays, if you want to reproduce
+    the measured delays between messages look at the
+    :class:`can.util.MessageSync` class.
+    """
+
+    @classmethod
+    def __new__(cls, other, filename):
+        if filename.endswith(".asc"):
+            raise NotImplemented
+        #     return ASCReader(filename)
+        if filename.endswith(".csv"):
+            raise NotImplemented
+        #     return CSVReader(filename)
+        if filename.endswith(".db"):
+            return SqlReader(filename)
+
+
+class SqlReader:
+    def __init__(self, filename):
+        log.debug("Starting sqlreader with {}".format(filename))
+        conn = sqlite3.connect(filename)
+
+        self.c = conn.cursor()
+        self.c.execute("SELECT ts FROM messages LIMIT 1")
+
+        self.recorded_start_time = self.c.fetchone()[0]
+
+        self.recorded_start_datetime = datetime.fromtimestamp(self.recorded_start_time)
+
+    @staticmethod
+    def create_frame_from_db_tuple(frame_data):
+        ts, id, is_extended, is_remote, is_error, dlc, data = frame_data
+        return Message(
+            ts, is_remote, is_extended, is_error, id, dlc, data
+        )
+
+    def __iter__(self):
```

Comment:
I think that we can rely on the garbage collector - especially because we aren't writing to the db here.